### PR TITLE
Various module improvements

### PR DESCRIPTION
**Pull request by RussellALA**
These changes include a few stability improvements and minor bugfixes to FrEIA modules:
- The hard permutation in the AllInOne coupling block implementation was changed from a permutation matrix multiplication (ch x ch parameters) to an indexing tensor (ch parameters), which significantly reduces model size especially after flattening high dimensional data
- softclamping in the AllInOne block now no longer scales the near-linear part by the clamping boundaries (clamp * tanh(x) -> clamp * tanh(x/clamp))
- added a new parameter "domain_clamping" to BinnedSplineBase, which softclamps the domain width and height of the spline to (-domain_clamping, domain_clamping), since total width/height > 0 is always true before clamping, effectively this restricts the splines to a domain of (0, domain_clamping).
- In rational quadratic splines, if the discriminant check fails, added the violating value to the error message to differentiate between numerical errors (negative discriminant close to 0) and training instabilities (NaN values).
- fixed a bug in ActNorm that prevented loading of >1D ActNorm layers. The actnorm scale and mean are now initialized to be full-dimensional, not just on the channel dimension (which only was the case after the values were set on the first batch).

### CODE REVIEW

File: FrEIA/modules/all_in_one_block.py [hunk 1]
```diff
@@ -143,6 +146,7 @@ class or callable ``f``, called as ``f(channels_in, channels_out)`` and
         if permute_soft:
             w = special_ortho_group.rvs(channels)
         else:
+            w_index = torch.randperm(channels, requires_grad=False)
             w = np.zeros((channels, channels))
             for i, j in enumerate(np.random.permutation(channels)):
                 w[i, j] = 1.
```
Comment (fdraxler) [inline FrEIA/modules/all_in_one_block.py:152]: Is this still needed?

File: FrEIA/modules/all_in_one_block.py [hunk 2]
```diff
@@ -155,11 +159,14 @@ class or callable ``f``, called as ``f(channels_in, channels_out)`` and
             self.w_perm = None
             self.w_perm_inv = None
             self.w_0 = nn.Parameter(torch.FloatTensor(w), requires_grad=False)
-        else:
+        elif permute_soft:
             self.w_perm = nn.Parameter(torch.FloatTensor(w).view(channels, channels, *([1] * self.input_rank)),
                                        requires_grad=False)
             self.w_perm_inv = nn.Parameter(torch.FloatTensor(w.T).view(channels, channels, *([1] * self.input_rank)),
                                            requires_grad=False)
```
Comment (fdraxler) [inline FrEIA/modules/all_in_one_block.py:166]: While we're here: Isn't `torch.from_numpy` preferred over tensor(array)?

File: FrEIA/modules/splines/binned.py [hunk 3]
```diff
@@ -75,6 +75,7 @@ def __init__(self, dims_in, dims_c=None, bins: int = 10, parameter_counts: Dict[
             default_domain: tuple of (left, right, bottom, top) default spline domain values
                 these values will be used as the starting domain (when the network outputs zero)
             identity_tails: whether to use identity tails for the spline
+            domain_clamping: clamping value for the domain
```
Comment (fdraxler) [inline FrEIA/modules/splines/binned.py:78]: Docstring could be more explicit -- how about "if float, clamp spline domain to (-domain_clamping, domain_clamping)"

File: FrEIA/modules/splines/binned.py [hunk 4]
```diff
@@ -161,7 +170,16 @@ def constrain_parameters(self, parameters: Dict[str, torch.Tensor]) -> Dict[str,
...
+
+            domain_width = torch.sum(parameters["widths"], dim=-1, keepdim=True)
+            domain_height = torch.sum(parameters["heights"], dim=-1, keepdim=True)
+            width_resize = self.clamp_domain(domain_width) / domain_width
+            height_resize = self.clamp_domain(domain_height) / domain_height
+
+            parameters["widths"] = parameters["widths"] * width_resize
+            parameters["heights"] = parameters["heights"] * height_resize
+            parameters["left"] = parameters["left"] * width_resize
+            parameters["bottom"] = parameters["bottom"] * height_resize
```
Comment (fdraxler) [inline FrEIA/modules/splines/binned.py:182]: I am not exactly sure about the internal behavior of the splines, maybe @LarsKue can comment on this :)

**Review by fdraxler — COMMENTED** [review]
Looks good, only minor comments :)

**Comment by fdraxler** [comment]
And it looks like one of the tests is not passing – can you run on your machine to see if this is caused by your changes?

File: FrEIA/modules/all_in_one_block.py (hunk 2 above)
Comment (LarsKue) [reply FrEIA/modules/all_in_one_block.py:166]: Also a bit strange that the parameter is being saved as a view. A `reshape` or `view().contiguous()` seems more appropriate, unless the `view` has a specific purpose?

File: FrEIA/modules/splines/binned.py [hunk 5]
```diff
@@ -92,6 +93,13 @@ def __init__(self, dims_in, dims_c=None, bins: int = 10, parameter_counts: Dict[
         assert default_domain[3] - default_domain[2] >= min_bin_sizes[1] * bins, \
         "{bins} bins of size {min_bin_sizes[1]} are too large for domain {default_domain[2]} to {default_domain[3]}"
 
+        if domain_clamping is not None:
+            self.clamp_domain = lambda domain: domain_clamping * torch.tanh(
+                domain / domain_clamping
+            )
+        else:
+            self.clamp_domain = lambda domain: domain
+
```
Comment (LarsKue) [inline FrEIA/modules/splines/binned.py:103]: I think we'd be better off making this a method, rather than a named lambda (which are discouraged by PEP-8)

File: FrEIA/modules/splines/rational_quadratic.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def rational_quadratic_spline(x: torch.Tensor,
 
         # Eq 29 in the appendix of the paper
         discriminant = b ** 2 - 4 * a * c
-        assert torch.all(discriminant >= 0)
+        assert torch.all(discriminant >= 0), f"Discriminant must be positive, but is violated by {torch.min(discriminant)}"
```
Comment (LarsKue) [inline FrEIA/modules/splines/rational_quadratic.py:167]: This was an assert, because I assumed it could never be violated. If we want error messages, we should probably `raise` and Exception.

File: FrEIA/modules/splines/binned.py (hunk 4 above)
Comment (LarsKue) [reply FrEIA/modules/splines/binned.py:183]: Looks good to me. Did log scaling the domain not work @RussellALA ?

File: FrEIA/modules/all_in_one_block.py (hunk 1 above)
Comment (RussellALA) [reply FrEIA/modules/all_in_one_block.py]: No, removed :)

File: FrEIA/modules/splines/binned.py (hunk 4 above)
Comment (RussellALA) [reply FrEIA/modules/splines/binned.py:183]: > Looks good to me. Did log scaling the domain not work @RussellALA ?
My test run ended in NaN. I might try a few more experiments, but it seems to not address the stability issues

File: FrEIA/modules/splines/rational_quadratic.py (hunk 6 above)
Comment (RussellALA) [reply FrEIA/modules/splines/rational_quadratic.py:167]: Fair point. This was actually just a diagnostics tool for me, to see what the cause of model the assertion was. But since the intended purpose of the assert it was to catch negative discriminants how about changing the line to:
```Python
assert not torch.any(discriminant < 0)
```
If the training diverges to NaN this should result in an error elsewhere.

Comment (LarsKue) [reply FrEIA/modules/splines/rational_quadratic.py:167]: > If the training diverges to NaN this should result in an error elsewhere.

While this is true, I think the current assertion is better, because it ensures the remaining algorithm is computable.

Comment (RussellALA) [reply FrEIA/modules/splines/rational_quadratic.py]: Agreed. I have changed it to a RuntimeError

File: FrEIA/modules/all_in_one_block.py (hunk 2 above)
Comment (RussellALA) [reply FrEIA/modules/all_in_one_block.py]: I tested the suggested changes, and they don't seem to violate any of my local or PR pipepline testing. `torch.from_numpy` loses the explicit typecast to float, and instead uses double (as the scipy default). I personally think this is unnecessary and if you agree I propose revert this change to `torch.FloatTensor`. as for `view().contiguous()` I see no reason this should not be included.

**Comment by RussellALA** [comment]
> And it looks like one of the tests is not passing – can you run on your machine to see if this is caused by your changes?

Yes, the AIO block failed to throw an Exception if the wrong input shape was given with hard permutes. $\rightarrow$ Fixed

**Comment by mjack3** [comment]
The new ActNorm implementation is not according to the official implementation.

**Comment by RussellALA** [comment]
> The new ActNorm implementation is not according to the official implementation.

In what respect do they differ? Note that the "new" implementation does not change how ActNorm functionally behaves in FrEIA, it simply fixes a bug that prevented non vector-valued models using ActNorm being loaded.

**Comment by mjack3** [comment]
In the latest pull Request, the ActNorm layer was modified by adding parameters per channel, height, and width. However, if we attend to the original paper:

"We propose an actnorm layer (for activation normalizaton), that performs an affine transformation of the activations using a scale and bias parameter per channel, similar to batch normalization".

This can also be checked using its official code. To sum up, to me the previous implementation was the correct one.

**Comment by LarsKue** [comment]
> In the latest pull Request, the ActNorm layer was modified by adding parameters per channel, height, and width

I verified this as true.

> The previous implementation was the correct one.

This is not correct. The later dimensions should be initialized to be of unit size. I think something like this code should solve the issue:

```python3
dims = next(iter(dims_in))
param_dims = copy(dims)
param_dims[1:] = 1
scale = torch.empty(1, *param_dims)
```

**Comment by mjack3** [comment]
And what is the problem regarding the original done by @ardizzone ?

```
self.dims_in = dims_in[0]
param_dims = [1, self.dims_in[0]] + [1 for i in range(len(self.dims_in) - 1)]
self.scale = nn.Parameter(torch.zeros(*param_dims))
self.bias = nn.Parameter(torch.zeros(*param_dims))
```
I see it perfectly well.

**Comment by LarsKue** [comment]
The original code from Lynton Ardizzone, while functional, was unreadable and not maintainable. I refactored this block to reflect its intended use in f1e73d1 and e0287a2, but my implementation relies on implicit tensor broadcasting. I am not sure when this is a problem, but nevertheless @RussellALA approached me later stating that it is. In any case, for the initialization, Lyntons code should work fine, and so should my proposal above.

I will push a solution to this on Monday, if @RussellALA has not by then, but otherwise I consider this issue solved.

**Comment by mjack3** [comment]
Well, although you did a good change, ActNorm should normalize along the channels only...am I wrong?

**Comment by RussellALA** [comment]
I think there has been some confusion what the actual change introduced in this PR was and how ActNorm should work. 


> In the latest pull Request, the ActNorm layer was modified by adding parameters per channel, height, and width.

This is incorrect. The ActNorm layer in FrEIA does indeed act per c x h x w, but was already modified in https://github.com/vislearn/FrEIA/commit/f1e73d1e60b3625a56d402c2d4f8089917947269 to do so.

> "We propose an actnorm layer (for activation normalizaton), that performs an affine transformation of the activations using a scale and bias parameter per channel, similar to batch normalization".

This is correct. ActNorm was not faithful to the original implementation in FrEIA and still is not. This should be fixed (in another PR ;) ).

The reason for the confusion might be that the parameters for actnorm were overwritten during initialization with the first batch. So event though we define 
```Python
dim = next(iter(dims_in))[0]
self.log_scale = nn.Parameter(torch.empty(1, dim))
self.loc = nn.Parameter(torch.empty(1, dim))
```
in `__init__`, later we call 
```Python
def initialize(self, batch: torch.Tensor):
    self.is_initialized.data = torch.tensor(True)
    self.log_scale.data = torch.log(torch.std(batch, dim=0, keepdim=True))
    self.loc.data = torch.mean(batch, dim=0, keepdim=True)
```
This overwrites the tensor data with a different shape than they were initialized with.
